data/multiplayer/scenarios/4p_King_of_the_Hill.cfg

```diff
@@ -4,20 +4,44 @@
     [store_side]
         side={SIDE}
     [/store_side]
-    {RANDOM $side.recruit}
-    [store_unit_type]
-        type=$random
-    [/store_unit_type]
-    # If the unit is level 0 then switch to one of its advancements instead.
-    [if]
-        [variable]
-            name=unit_type.level
-            equals=0
-        [/variable]
-        [then]
-            {RANDOM $unit_type.advances_to}
-        [/then]
-    [/if]
+    # Get the faction leader list of the side.
+    [lua]
+        code=<<
+            local t = ...
+            local era = wesnoth.get_era(wesnoth.scenario.era.id)
+            for i,v in ipairs(era) do
+                if v["tag"] == "multiplayer_side" then
+                    for k,faction in pairs(v) do
+                        if faction["id"] == t.faction then
+                            wesnoth.set_variable("random", faction["leader"])
+                        end
+                    end
+                end
+            end
+            >>
+        [args]
+            faction=$side.faction
+        [/args]
+    [/lua]
+    # Pick a random leader from the list and then switch to a level 1 version of it.
+    {RANDOM $random}
+    [lua]
+        code=<<
+            local t = ...
+            for id,unit_type in pairs(wesnoth.unit_types) do
+                for i,advancement in pairs(unit_type["advances_to"]) do
+                    if advancement == t.advancement and wesnoth.unit_types[id]["level"] == 1 then
```

Comment:
I think this logic isn't going to work on an Age of Heroes era – there will be no level 1 unit found. However, in Age of Heroes I think I would expect it to choose a level 2 version of the level 3 leader, so why not just check that the level is less than the level of the main unit? ie, `wesnoth.unit_types[id].level < wesnoth.unit_types[t.advancement].level` or so.

If you _really_ want it to be a level 1 even in Age of Heroes, you're going to have to make this more complicated, because it'll have to find the level 2 that advances to it, and then the level 1 that advances to that level 2.

There's also the question of whether to do something special to handle converging unit lines. It doesn't seem like a high priority though.

Reply:
There was (and maybe still is) a rare bug where instead of being "demoted" one level, the heir would be demoted two or more levels somehow. So in Default some players would rarely get a level 0 heir or in Age of Heroes a level 1 heir. Which lead me to assume that `advances_to` was somehow walking all levels of the unit advancement line. So I tried to lock it to level 1 for consistency (plus for some other balance reasons).

But it looks like the level lock did break the demotion mechanic in Age of Heroes as you predicted, so I removed it.

Reply:
We decided to remove this whole "demotion" mechanic to support UMC eras that have leaders with no level 1 in their advancement line. Now second leaders are simply pulled from the faction leader list and used as is.